File: modules/prop-shop/src/trading/multi_strategy/strategy_registry.py
"""Strategy Registry - Maintains validated strategies with performance metrics."""

import json
from pathlib import Path
from typing import Dict, List, Optional
from datetime import datetime

class StrategyRegistry:
    """Registry for validated trading strategies with live performance tracking."""
    
    def __init__(self, backtest_results_path: str):
        """Initialize registry from backtest results JSON."""
        self.strategies: Dict = {}
        self.live_stats: Dict[str, Dict] = {}
        self.backtest_results_path = backtest_results_path
        self._load_strategies()
    
    def _load_strategies(self):
        """Load validated strategies from backtest results."""
        with open(self.backtest_results_path, 'r') as f:
            data = json.load(f)
        
        strategies_data = data.get('strategies', {})
        
        # Map strategy names to their combined stats
        strategy_mapping = {
            'rsi_oversold': 'RSI < 40',
            'rsi_macd': 'RSI+MACD',
            'rsi_bollinger_bands': 'RSI+BB',
        }
        
        for key, backtest_name in strategy_mapping.items():
            if backtest_name in strategies_data:
                combined = strategies_data[backtest_name].get('combined', {})
                self.strategies[key] = {
                    'name': key,
                    'display_name': backtest_name,
                    'validated': True,
                    'backtest_stats': {
                        'win_rate': combined.get('win_rate', 0),
                        'profit_factor': combined.get('profit_factor', 0),
                        'sharpe_ratio': combined.get('sharpe_ratio', 0),
                        'trade_count': combined.get('trade_count', 0),
                    },
                    'entry_rule_name': self._get_entry_rule(key),
                }
                
                # Initialize live stats
                self.live_stats[key] = {
                    'trades': 0,
                    'wins': 0,
                    'losses': 0,
                    'live_win_rate': 0.0,
                    'active': True,
                }
    
    def _get_entry_rule(self, strategy_key: str) -> str:
        """Get entry rule description for strategy."""
        rules = {
            'rsi_oversold': 'RSI < 40',
            'rsi_macd': 'RSI < 40 AND MACD histogram positive',
            'rsi_bollinger_bands': 'RSI < 40 AND price at lower Bollinger Band',
        }
        return rules.get(strategy_key, 'Unknown')
    
    def register_trade(self, strategy_key: str, won: bool) -> None:
        """Update live stats for a strategy after trade completion."""
        if strategy_key not in self.live_stats:
            return
        
        stats = self.live_stats[strategy_key]
        stats['trades'] += 1
        
        if won:
            stats['wins'] += 1
        else:
            stats['losses'] += 1
        
        # Update live win rate
        if stats['trades'] > 0:
            stats['live_win_rate'] = (stats['wins'] / stats['trades']) * 100
        
        # Check edge decay: retire if WR < 40% after 10+ trades
        if stats['trades'] >= 10 and stats['live_win_rate'] < 40:
            stats['active'] = False
    
    def retire_strategy(self, strategy_key: str) -> None:
        """Retire a strategy due to edge decay."""
        if strategy_key in self.live_stats:
            self.live_stats[strategy_key]['active'] = False
    
    def get_active_strategies(self) -> List[str]:
        """Get list of active strategy keys."""
        return [
            key for key, stats in self.live_stats.items()
            if stats['active'] and key in self.strategies
        ]
    
    def get_strategy(self, strategy_key: str) -> Optional[Dict]:
        """Get strategy details."""
        return self.strategies.get(strategy_key)
    
    def get_live_stats(self, strategy_key: str) -> Optional[Dict]:
        """Get live performance stats for a strategy."""
        return self.live_stats.get(strategy_key)
    
    def get_all_strategies_status(self) -> Dict:
        """Get status of all strategies."""
        status = {}
        for key in self.strategies:
            stats = self.live_stats[key]
            strategy = self.strategies[key]
            status[key] = {
                'display_name': strategy['display_name'],
                'backtest_wr': strategy['backtest_stats']['win_rate'],
                'live_wr': stats['live_win_rate'],
                'trades': stats['trades'],
                'active': stats['active'],
            }
        return status

File: src/trading/strategy-registry.ts
import { readFileSync } from "fs";
import { join } from "path";

export interface BacktestStats {
  win_rate: number;       // 0–100 (as stored in JSON)
  profit_factor: number;
  sharpe_ratio: number;
  trade_count: number;
}

export interface Strategy {
  name: string;
  description: string;
  validated: boolean;
  backtest: BacktestStats;
  live_trades: number;
  live_wins: number;
  retired: boolean;
}

/** Minimum live trades before retirement kicks in */
const RETIREMENT_THRESHOLD_TRADES = 10;
/** Retire if live win rate drops below this (0–1) */
const RETIREMENT_MIN_WIN_RATE = 0.40;

export function liveWinRate(s: Strategy): number {
  return s.live_trades > 0 ? s.live_wins / s.live_trades : s.backtest.win_rate / 100;
}

/** Composite score: live WR (primary) + Sharpe normalised (secondary) + profit factor */
export function scoreStrategy(s: Strategy): number {
  const wr = liveWinRate(s);
  // Normalise Sharpe to a 0-1 contribution (cap at 10)
  const sharpeNorm = Math.min(Math.max(s.backtest.sharpe_ratio, 0), 10) / 10;
  return wr * 0.6 + sharpeNorm * 0.2 + Math.min(s.backtest.profit_factor, 5) / 5 * 0.2;
}

export class StrategyRegistry {
  private strategies = new Map<string, Strategy>();

  register(strategy: Strategy): void {
    this.strategies.set(strategy.name, { ...strategy });
  }

  getAll(): Strategy[] {
    return Array.from(this.strategies.values()).filter((s) => !s.retired);
  }

  getById(name: string): Strategy | undefined {
    return this.strategies.get(name);
  }

  /** Increment live stats and retire strategy if win rate decays too far. */
  registerTrade(name: string, won: boolean): void {
    const s = this.strategies.get(name);
    if (!s) throw new Error(`Strategy not found: ${name}`);
    s.live_trades += 1;
    if (won) s.live_wins += 1;

    const liveWR = liveWinRate(s);
    if (s.live_trades >= RETIREMENT_THRESHOLD_TRADES && liveWR < RETIREMENT_MIN_WIN_RATE) {
      s.retired = true;
      console.warn(
        `[StrategyRegistry] ${name} retired — live WR ${(liveWR * 100).toFixed(1)}% < ${RETIREMENT_MIN_WIN_RATE * 100}%`
      );
    }
  }

  rankStrategies(): Strategy[] {
    return this.getAll().sort((a, b) => scoreStrategy(b) - scoreStrategy(a));
  }

  /** Load strategies from backtest_results.json summary section. */
  static loadFromBacktestFile(filePath: string): StrategyRegistry {
    const raw = readFileSync(filePath, "utf-8");
    const data = JSON.parse(raw) as {
      summary: Record<string, { win_rate: number; profit_factor: number; sharpe_ratio: number; trade_count: number }>;
    };

    const registry = new StrategyRegistry();

    const nameMap: Record<string, { id: string; description: string }> = {
      "RSI < 40": {
        id: "rsi_oversold",
        description: "RSI oversold bounce (RSI < 40)",
      },
      "RSI+MACD": {
        id: "rsi_macd_confirmation",
        description: "RSI oversold + MACD bullish crossover confirmation",
      },
      "RSI+BB": {
        id: "rsi_bollinger_bands",
        description: "RSI oversold + price at lower Bollinger Band",
      },
    };

    for (const [jsonName, stats] of Object.entries(data.summary)) {
      const meta = nameMap[jsonName];
      if (!meta) continue;

      registry.register({
        name: meta.id,
        description: meta.description,
        validated: true,
        backtest: {
          win_rate: stats.win_rate,
          profit_factor: stats.profit_factor,
          sharpe_ratio: stats.sharpe_ratio,
          trade_count: stats.trade_count,
        },
        live_trades: 0,
        live_wins: 0,
        retired: false,
      });
    }

    return registry;
  }
}

// ── Singleton loaded from backtest results ──────────────────────────────────
const backtestPath = join(
  import.meta.dirname ?? process.cwd(),
  "..",
  "..",
  "projects",
  "prop-shop",
  "backtest_results.json"
);

export const strategyRegistry = StrategyRegistry.loadFromBacktestFile(backtestPath);
